core/templates/pages/exploration-player-page/services/stats-reporting.service.spec.ts

```diff
@@ -0,0 +1,294 @@
+    spyOn(urlService, 'getUrlParams')
+      .and.returnValue({classroom_url_fragment: 'classroom'});
+    StatsReportingService.stateStopwatch = Stopwatch.create();
+  });
+
+  it('should create default aggregated stats when initialized', () => {
+    let defaultValues: AggregatedStats = {
+      num_starts: 0,
+      num_completions: 0,
+      num_actual_starts: 0,
+      state_stats_mapping: {}
+    };
+
+    expect(StatsReportingService.aggregatedStats).toEqual(defaultValues);
+  });
+
+  it('should set session properties when calling ' +
+    '\'initSession\'', fakeAsync(() => {
+    statsReportingService.initSession(
+      explorationId, explorationTitle, explorationVersion,
+      sessionId, collectionId);
+    tick(300001);
+    discardPeriodicTasks();
+
+    expect(StatsReportingService.explorationId).toEqual(explorationId);
+    expect(StatsReportingService.explorationTitle).toEqual(explorationTitle);
+    expect(StatsReportingService.explorationVersion)
+      .toEqual(explorationVersion);
+    expect(StatsReportingService.sessionId).toEqual(sessionId);
+    expect(StatsReportingService.optionalCollectionId).toEqual(collectionId);
+  }));
+
+  it('should record exploration\'s stats when it is about to start', () => {
+    StatsReportingService.explorationStarted = false;
+    statsReportingService.recordExplorationStarted('firstState', {});
+
+    expect(StatsReportingService.explorationStarted).toBe(true);
+    expect(recordExplorationStartedSpy).toHaveBeenCalled();
+  });
+
+  it('should not again record exploration\'s stats when ' +
+    'it is about to start and already recorded', () => {
+    StatsReportingService.explorationStarted = true;
+    statsReportingService.recordExplorationStarted('firstState', {});
+
+    expect(recordExplorationStartedSpy).not.toHaveBeenCalled();
+  });
+
+  it('should not create default stats for a state if ' +
+    'it\'s already exist', () => {
+    let defaultStats = {
+      total_answers_count: 0,
+      useful_feedback_count: 0,
+      total_hit_count: 0,
+      first_hit_count: 0,
+      num_times_solution_viewed: 0,
+      num_completions: 0
+    };
+
+    let originalStats = {
+      total_answers_count: 1,
+      useful_feedback_count: 1,
+      total_hit_count: 1,
+      first_hit_count: 1,
+      num_times_solution_viewed: 1,
+      num_completions: 1
+    };
+
+    StatsReportingService.aggregatedStats.state_stats_mapping = {
+      firstState: originalStats
+    };
+
+    // Trying to start a state which has already been recorded.
+    statsReportingService.recordExplorationStarted('firstState', {});
+
+    expect(StatsReportingService.aggregatedStats.state_stats_mapping.firstState)
+      .toBe(originalStats);
+    expect(StatsReportingService.aggregatedStats.state_stats_mapping.firstState)
+      .not.toBe(defaultStats);
+  });
+
+  it('should record exploration\'s stats when it is actually started', () => {
+    StatsReportingService.explorationActuallyStarted = false;
```

Comment:
@nithusha21 your suggestion did the trick. But, both `StatsReportingService` and `statsReportingService` are not updating simultaneously in the tests.
If I place prechecks at the beginning of the test they are flaking (almost 90%  of the cases).
So in order to prevent that I've set the testing value to be null for each test.
(I feel that the ideal thing to do is change the static variable to public.)

Reply:
PTAL @nithusha21

Reply:
I think I like your suggestion to make it public. Do you know why it is static? In general, if the code feels off, feel free to make it correct.

Reply:
I don't think we really need static here.
As it is not being used in hybrid state or in common anywhere else.
I'll try making them public and see if anything breaks.